pkg/writer.go

```diff
@@ -107,6 +142,108 @@ func (s *writerImpl) Close() (err error) {
 	return
 }
 
+type encodeResult struct {
+	buf   []byte
+	entry seekTableEntry
+}
+
+func (s *writerImpl) writeManyEncoder(ctx context.Context, ch chan<- encodeResult, frame []byte) func() error {
+	return func() error {
+		dst, entry, err := s.encodeOne(frame)
+		if err != nil {
+			return err
+		}
+
+		select {
+		case <-ctx.Done():
+		// Fulfill our promise
+		case ch <- encodeResult{dst, entry}:
+			close(ch)
+		}
+
+		return nil
+	}
+}
+
+func (s *writerImpl) writeManyReader(ctx context.Context, frameSource FrameSource, g *errgroup.Group, queue chan<- chan encodeResult) func() error {
+	return func() error {
+		for {
+			frame, err := frameSource()
+			if err != nil {
+				return err
+			}
+			if frame == nil {
+				close(queue)
+				return nil
+			}
+
+			// Put a channel on the queue as a sort of promise.
+			// This is a nice trick to keep our results ordered, even when compression
+			// completes out-of-order.
+			ch := make(chan encodeResult)
+			select {
+			case <-ctx.Done():
+				return nil
+			case queue <- ch:
+			}
+
+			g.Go(s.writeManyEncoder(ctx, ch, frame))
+		}
+	}
+}
+
+func (s *writerImpl) writeManyWriter(ctx context.Context, callback func(uint32), queue <-chan chan encodeResult) func() error {
+	return func() error {
+		for {
+			var ch <-chan encodeResult
+			select {
+			case <-ctx.Done():
+				return nil
+			case ch = <-queue:
+			}
+			if ch == nil {
+				return nil
+			}
+
+			// Wait for the block to be complete
+			var result encodeResult
+			select {
+			case <-ctx.Done():
+				return nil
+			case result = <-ch:
```

Comment:
we'll need to check that there is actually a result here, otherwise when channel is closed due to the context cancellation/error, this would lead to a zero-valued result.

Reply:
This channel is only closed after we put something in it. Either it has a value, or it's stuck empty (and we'll eventually take the ctx.Done side due to error), it should never be closed here.

I guess I could add a check anyway? But it will be impossible to trigger.

Reply:
ok, fair point, if we only close the channel on the successful path we do not need this.